zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/factory/SharedZkClientFactory.java

```diff
@@ -117,6 +117,9 @@ private ZkConnectionManager getOrCreateZkConnectionManager(
   private void cleanupConnectionManager(ZkConnectionManager zkConnectionManager) {
     synchronized (_connectionManagerPool) {
       zkConnectionManager.close(true);
+      if (!zkConnectionManager.hasSharedWatchers()) {
```

Comment:
This shall work, but will it be easier if we let close() return a boolean to indicate if the real close has been done? If true, then it is safe to remove it from the pool.

It is both easier and safer because once a zkConnectionManager is closed, there is no way to reuse it.
So even if later we modify the concurrency control, there is no risk at all.

Reply:
Thanks for the feedback, will incorporate this